### PR TITLE
vscode: add syntax tree inspection hovers and highlights

### PR DESCRIPTION
**Pull request by Veetaha**
inspect-tree
I implemented the reverse mapping (when you hover in the rust editor), but it seems overcomplicated, so I removed it

Related #3682

### CODE REVIEW

**Comment by Veetaha** [comment]
I implemented the reverse mapping (when you hover the in the editor), but it seems overcomplicated

File: editors/code/src/commands/syntax_tree.ts [hunk 1]
```diff
@@ -48,45 +28,128 @@ export function syntaxTree(ctx: Ctx): Cmd {
...
-}
-
-
-class TextDocumentContentProvider implements vscode.TextDocumentContentProvider {
-    uri = vscode.Uri.parse('rust-analyzer://syntaxtree');
-    eventEmitter = new vscode.EventEmitter<vscode.Uri>();
+class TextDocumentContentProvider implements vscode.TextDocumentContentProvider, Disposable {
+    readonly uri = vscode.Uri.parse('rust-analyzer://syntaxtree');
+    readonly eventEmitter = new vscode.EventEmitter<vscode.Uri>();
+    private readonly disposables: Disposable[] = [];
```
Comment (matklad) [inline editors/code/src/commands/syntax_tree.ts]: tdcp is a singletop, so we can just re-use `ctx.subscriptions` without implementing our our disposing logic.

File: editors/code/src/commands/syntax_tree.ts [hunk 2]
```diff
@@ -48,45 +28,128 @@ export function syntaxTree(ctx: Ctx): Cmd {
...
+
+// FIXME: consider implementing this via the Tree View API?
+// https://code.visualstudio.com/api/extension-guides/tree-view
+class AstInspector implements vscode.HoverProvider, Disposable {
+    private static readonly astDecorationType = vscode.window.createTextEditorDecorationType({
+        fontStyle: "normal",
+        border: "#ffffff 1px solid",
+    });
+    private rustEditor: undefined | RustEditor;
+    private readonly disposables: Disposable[] = [];
```
Comment (matklad) [inline editors/code/src/commands/syntax_tree.ts]: same here, there's only ever one of these, so it doesn't need its own disposable array.

**Comment by Veetaha** [comment]
@matklad 
Shouldn't we use the unifrom `disposables` array for handles? Using `ctx.subscriptions` is the straightforward, but seems to  be *the wrong* way...

**Comment by matklad** [comment]
We can use existing thing, I see no reason not to use it.

**Comment by Veetaha** [comment]
Migrated to the existing thing, added docs.
Also, I should not on a caveat. This works only when you have a single rust file editor visible, otherwise, it may highlight the wrong file, but I don't this will be an issue since this is just a debugging convenience.

File: editors/code/src/commands/syntax_tree.ts [hunk 3]
```diff
@@ -48,45 +28,125 @@ export function syntaxTree(ctx: Ctx): Cmd {
...
+        vscode.window.onDidChangeVisibleTextEditors(this.onDidChangeVisibleTextEditors, this, ctx.subscriptions);
+
+        ctx.pushCleanup(this);
+    }
+    dispose() {
+        this.setRustEditor(undefined);
+    }
+
+    private onDidCloseTextDocument(doc: vscode.TextDocument) {
+        if (!!this.rustEditor && doc.uri.toString() === this.rustEditor.document.uri.toString()) {
```
Comment (Veetaha) [inline editors/code/src/commands/syntax_tree.ts]:
```suggestion
        if (this.rustEditor && doc.uri.toString() === this.rustEditor.document.uri.toString()) {
```

**Comment by lnicola** [comment]
This doesn't play nice with light themes :sweat_smile:. I was wondering why I don't see the border.

**Comment by Veetaha** [comment]
ggwp

**Pull request merged**